### PR TITLE
Add ScyllaDB Store

### PR DESCRIPTION
**Pull request by arnenissen**
@djkevincr @madhawa-gunasekara Please review the final version of the pull request.

### CODE REVIEW

**Comment by djkevincr** [comment]
@arnenissen Thank you for your hard work during the GSoC period. PR looks very good. I m going to keep PR for a while for let others to review. Good Luck with your final evaluation.

**Review by djkevincr — APPROVED** [review]

**Comment by djkevincr** [comment]
Please raise if there are any concerns. Otherwise I will proceed merging this PR to master.

File: gora-cassandra/src/main/java/org/apache/gora/cassandra/store/CassandraClient.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public Cluster getCluster() {
     return cluster;
   }
 
-  void initialize(Properties properties, CassandraMapping mapping) throws Exception {
+  public void initialize(Properties properties, CassandraMapping mapping) throws Exception {
```
Comment (lewismc) [inline gora-cassandra/src/main/java/org/apache/gora/cassandra/store/CassandraClient.java:88]: @arnenissen please remove all of this cassandra logic. This is orthogonal to the goal of this pull request. Thank you

File: gora-scylladb/pom.xml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    <parent>
+        <artifactId>gora</artifactId>
+        <groupId>org.apache.gora</groupId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>gora-scylladb</artifactId>
+    <packaging>bundle</packaging>
+
```
Comment (lewismc) [inline gora-scylladb/pom.xml:30]: Remove whitespace.

File: gora-scylladb/pom.xml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        <system>JIRA</system>
+        <url>https://issues.apache.org/jira/browse/GORA</url>
+    </issueManagement>
+    <ciManagement>
+        <system>Jenkins</system>
+        <url>https://builds.apache.org/job/Gora-trunk/</url>
+    </ciManagement>
+
+    <properties>
+        <guava.version>18.0</guava.version>
```
Comment (lewismc) [inline gora-scylladb/pom.xml:55]: Is this not defined in parent pom?

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @arnenissen this is a BIG pull request. Congratulations on getting this far. I will try to work with you to get your code into the Gora projecy.
Please see my initial comments... thanks.

File: gora-scylladb/pom.xml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.apache.maven.surefire</groupId>
+                        <artifactId>surefire-junit47</artifactId>
+                        <version>2.20</version>
```
Comment (kamaci) [inline gora-scylladb/pom.xml:102]: It is already defined at parent pom.

File: gora-scylladb/src/examples/avro/scyllaDBKey.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
+{
+  "type": "record",
+  "name": "ScyllaDBKey",
+  "default": null,
+  "namespace": "org.apache.gora.scylladb.example.generated.AvroSerialization",
+  "doc": "This Object is created to used as Cassandra Key to test cassandra data store, Cassandra Key can be used to define partition keys, clustering keys.",
```
Comment (kamaci) [inline gora-scylladb/src/examples/avro/scyllaDBKey.json:6]: Cassandra?

File: gora-scylladb/src/examples/avro/scyllaDBRecord.json [hunk 6]
```diff
@@ -0,0 +1,107 @@
+{
+  "type": "record",
+  "name": "ScyllaDBRecord",
+  "default": null,
+  "namespace": "org.apache.gora.scylladb.example.generated.AvroSerialization",
+  "doc": "This object created to used as Persistent Object to test cassandra data store",
```
Comment (kamaci) [inline gora-scylladb/src/examples/avro/scyllaDBRecord.json:6]: Cassandra?

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+ * 
+ *Unless required by applicable law or agreed to in writing, software
+ *distributed under the License is distributed on an "AS IS" BASIS,
+ *WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *See the License for the specific language governing permissions and
+ *limitations under the License.
+ */
+package org.apache.gora.scylladb.example.generated.AvroSerialization;  
+
+/** This Object is created to used as Cassandra Key to test cassandra data store, Cassandra Key can be used to define partition keys, clustering keys. */
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java:20]: Cassandra?

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    return isDirty(1);
+  }
+
+  /** Creates a new ScyllaDBKey RecordBuilder */
+  public static org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder newBuilder() {
+    return new org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder();
+  }
+  
+  /** Creates a new ScyllaDBKey RecordBuilder by copying an existing Builder */
+  public static org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder newBuilder(org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder other) {
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java:154]: Line length is too much. Please check all the PR.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    return new org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder(other);
+  }
+  
+  /** Creates a new ScyllaDBKey RecordBuilder by copying an existing ScyllaDBKey instance */
+  public static org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey.Builder newBuilder(org.apache.gora.scylladb.example.generated.AvroSerialization.ScyllaDBKey other) {
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java:159]: Line length is too much.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    }
+  
+    /**
+     * Checks the dirty status of the 'timestamp' field. A field is dirty if it represents a change that has not yet been written to the database.
+         * @param value the value to set.
+     */
+    public boolean isTimestampDirty() {
+      throw new java.lang.UnsupportedOperationException("IsDirty is not supported on tombstones");
+    }
+
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBKey.java:344]: Remove extra lines.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java [hunk 11]
```diff
@@ -0,0 +1,1299 @@
...
+    DATA_DOUBLE(3, "dataDouble"),
+    DATA_BYTES(4, "dataBytes"),
+    ARRAY_INT(5, "arrayInt"),
+    ARRAY_STRING(6, "arrayString"),
+    ARRAY_LONG(7, "arrayLong"),
+    ARRAY_DOUBLE(8, "arrayDouble"),
+    MAP_INT(9, "mapInt"),
+    MAP_STRING(10, "mapString"),
+    MAP_LONG(11, "mapLong"),
+    MAP_DOUBLE(12, "mapDouble"),
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java:38]: No need a comma before a semicolon.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java [hunk 12]
```diff
@@ -0,0 +1,1299 @@
...
+     * Field's name.
+     */
+    private String name;
+
+    /**
+     * Field's constructor
+     * @param index field's index.
+     * @param name field's name.
+     */
+    Field(int index, String name) {this.index=index;this.name=name;}
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java:55]: Start the method body with a new line.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java [hunk 13]
```diff
@@ -0,0 +1,1299 @@
...
+
+    /**
+     * Gets field's attributes to string.
+     * @return String field's attributes to string.
+     */
+    public String toString() {return name;}
+  };
+
+  public static final String[] _ALL_FIELDS = {
+  "dataString",
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java:77]: Check the indentation.

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java [hunk 14]
```diff
@@ -0,0 +1,1299 @@
...
+
+  /**
+   * Gets the total field count.
+   * @return int field count
+   */
+  public int getFieldsCount() {
+    return ScyllaDBRecord._ALL_FIELDS.length;
+  }
+
+  private java.lang.CharSequence dataString;
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java:100]: java.lang is imported as default

File: gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java [hunk 15]
```diff
@@ -0,0 +1,1299 @@
...
+  // Used by DatumReader.  Applications should not call. 
+  @SuppressWarnings(value="unchecked")
+  public void put(int field$, java.lang.Object value) {
+    switch (field$) {
+    case 0: this.dataString = (java.lang.CharSequence)(value); break;
+    case 1: this.dataInt = (java.lang.Integer)(value); break;
+    case 2: this.dataLong = (java.lang.Long)(value); break;
+    case 3: this.dataDouble = (java.lang.Double)(value); break;
+    case 4: this.dataBytes = (java.nio.ByteBuffer)(value); break;
+    case 5: this.arrayInt = (java.util.List<java.lang.Integer>)((value instanceof org.apache.gora.persistency.Dirtyable) ? value : new org.apache.gora.persistency.impl.DirtyListWrapper((java.util.List)value)); break;
```
Comment (kamaci) [inline gora-scylladb/src/examples/java/org/apache/gora/scylladb/example/generated/AvroSerialization/ScyllaDBRecord.java:143]: Check all the line lengths here.

**Comment by kamaci** [comment]
Thanks for your effort @arnenissen? Could you explain what is the difference using Cassandra client of Gora instead of this since ScyllaDB is designed to be compatible with Cassandra? 

Can anybody use current Cassandra client to interact with ScyllaDB or not?
